[c10d][libuv] add partial read test for libuv backend and fix an error which only happens when partially reading a buffer

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116180
* __->__ #116141

**Test Plan**
1. build pytorch
2. execute `TORCH_CPP_LOG_LEVEL=INFO build/bin/TCPStoreTest --gtest_filter=TCPStoreTest.testLibUVPartialRead` from the pytorch root directory.

without the change:
<img width="761" alt="image" src="https://github.com/pytorch/pytorch/assets/12968408/1942e3c2-a9c1-4fe4-87e8-7e21f4d8f9aa">


with the change:
<img width="747" alt="image" src="https://github.com/pytorch/pytorch/assets/12968408/f3e96a5b-0ed1-49bd-9184-bb8a5ebebc33">


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

XilunWu committed:
[c10d][libuv] add partial read test for libuv backend and fix an error which only happens when partially reading a buffer

[ghstack-poisoned]

XilunWu committed:
Update on "[c10d][libuv] add partial read test for libuv backend and fix an error which only happens when partially reading a buffer"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu committed:
Update on "[c10d][libuv] add partial read test for libuv backend and fix an error which only happens when partially reading a buffer"


**Test Plan**
1. build pytorch
2. execute `TORCH_CPP_LOG_LEVEL=INFO build/bin/TCPStoreTest --gtest_filter=TCPStoreTest.testLibUVPartialRead` from the pytorch root directory.

without the change:
<img width="761" alt="image" src="https://github.com/pytorch/pytorch/assets/12968408/1942e3c2-a9c1-4fe4-87e8-7e21f4d8f9aa">


with the change:
<img width="747" alt="image" src="https://github.com/pytorch/pytorch/assets/12968408/f3e96a5b-0ed1-49bd-9184-bb8a5ebebc33">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -396,6 +396,18 @@ void TCPStore::validate(void) {
   buffer.flush();
 }
 
+void TCPStore::_splitSet(
```
Can we actually make this into test by inherit TCPStore and add it to the child class instead of adding it to TCPStore itself?

XilunWu replied:
IMO that would require a refactor of TCPStore because the TCPClient object is a Private member of TCPStore and can only be accessed from the TCPStore object.

wconstab replied:
i think @fduwjj means, can you make a TestTCPStore class in TcpStoreTest.cpp, and just implement the extra method on that?

XilunWu replied:
I have tried that attempt, but the thing is, the subclass cannot access the private member of the superclass, so the _splitSet function in the subclass won't be able to use the buffer

wconstab replied:
ok. could probably change the base class member to protected, add your test class as a friend class, or just copy some code into the test class.

Or, maybe you don't even need a tcpstore at all- just create a raw tcpclient obj in your test case, and then write the 'split_set' as a standalone helper function that uses tcpclient?

XilunWu replied:
Totally agree with the second solution (do not create a TCPStore object in the client thread but directly copy paste the socket connect and send logic). But I would prefer to merging this PR first then make that change in a follow-up to speed up since it's not harming. WDYT? @wconstab @fduwjj

wconstab replied:
ok. i was thinking about saying that you should just reverse the order, land the next refactor PR first, then land this one.  But i think its worth landing the fix now and its not that bad to add the _split_set, as long as you promise to clean it up soon.

XilunWu replied:
@wconstab @fduwjj for the follow-up there're several options I can come up with, let me know your opinions:
1. in test's TCP store client thread, do not use any `TCPStore`, `SendBuffer`, or `TCPClient` because they either cannot be used outside of `TCPStore.cpp` or encapsulate private members from users. Instead, use the `Socket` to perform the exact same thing as in those classes. The downside is making the test quite verbose (but we can hide them in test util files).

2. change some members to Protected and make the test class a friend class. The downside is I don't know the consequence of making the underlying networking components `Protected`. Can there be any security issue? Who would be the expert we can ask?

3. Refactor the structure of `TCPStore.hpp` and `TCPStoreBackend.hpp`. Currently, `TCPClient` and `SendBuffer` can only have incomplete declaration in `TCPStore.hpp` because `QueryType` is defined in `TCPStoreBackend.hpp` and `TCPStoreBackend.hpp` include `TCPStore.hpp`. This forbids them from being exposed outside of `TCPStore.cpp`. The question is, do we want to refactor them? Any benefit? A single test may not be a good justification.

fduwjj replied:
I think #2 should be fine? What security issue can you think of?

XilunWu replied:
If another class A inherits `TCPStore`, the class A or an instance of A can access a `TCPStore`'s buffer and read/write on it. I assume the original design decision (making them `private`) is to avoid this.

fduwjj replied:
Well, what's the problem for that? If it's a child of `TCPStore` then it's ok to access the buffer?

fduwjj commented on test/cpp/c10d/TCPStoreTest.cpp:
```diff
@@ -204,6 +205,50 @@ TEST(TCPStoreTest, testCleanShutdown) {
 }
 
+TEST(TCPStoreTest, testLibUVPartialRead) {
+  int numWorkers = 2;
```
Wait, we say two clients here but we only have one client in the test?

XilunWu replied:
Oops, that's a typo. Since we did not wait for workers, this was not captured in test.... 🙃

fduwjj replied:
I guess what I am trying to say is that you basically only use one client not two. If so why do you even need one extra thread (aka, extra client) if you only use one?

XilunWu replied:
Oh I see. That is right, we didn't utilize the client created in the main thread. Good point! I think you're right we can remove the extra thread.

fduwjj commented on test/cpp/c10d/TCPStoreTest.cpp:
```diff
@@ -204,6 +205,50 @@ TEST(TCPStoreTest, testCleanShutdown) {
+      c10d::test::check(*clientTCPStore, key, "v");
+    }
+  });
+
```
Do you also need `serverTCPStore = nullptr;` here?

XilunWu replied:
That will cause the issue. If we have it here, the server could be destructed before the client completes its execution.

fduwjj replied:
Then maybe add after clientThread Join?

XilunWu replied:
When the `unique_ptr` object `serverTCPStore` goes out of the scope (in this case it's the end of the test), the object will be destroyed (same as `serverTCPStore = nullptr`) and the TCPStore will be freed. So adding `serverTCPStore = nullptr` will just be the same as the current code.

fduwjj replied:
I guess my point is to set to null after join, by the time of join, all clients should have finished? But since this is a unique pointer, it should be fine.

XilunWu commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -396,6 +396,18 @@ void TCPStore::validate(void) {
+  detail::SendBuffer buffer(*client_, detail::QueryType::SET);
+  buffer.appendString(keyPrefix_ + key);
+  buffer.flush();
+  sleep(1);
```
switch to a platform neutral function `std::this_thread::sleep_for`.

XilunWu committed:
Update on "[c10d][libuv] add partial read test for libuv backend and fix an error which only happens when partially reading a buffer"


**Test Plan**
1. build pytorch
2. execute `TORCH_CPP_LOG_LEVEL=INFO build/bin/TCPStoreTest --gtest_filter=TCPStoreTest.testLibUVPartialRead` from the pytorch root directory.

without the change:
<img width="761" alt="image" src="https://github.com/pytorch/pytorch/assets/12968408/1942e3c2-a9c1-4fe4-87e8-7e21f4d8f9aa">


with the change:
<img width="747" alt="image" src="https://github.com/pytorch/pytorch/assets/12968408/f3e96a5b-0ed1-49bd-9184-bb8a5ebebc33">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge
